cpp/src/arrow/tensor_benchmark.cc

```diff
@@ -0,0 +1,67 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/record_batch.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/type.h"
+#include "arrow/util/benchmark_util.h"
+
+namespace arrow {
+
+template <typename ValueType>
+static void BatchToTensorSimple(benchmark::State& state) {
+  using CType = typename ValueType::c_type;
+  std::shared_ptr<DataType> ty = TypeTraits<ValueType>::type_singleton();
+
+  const int64_t num_cols = state.range(1);
+  const int64_t num_rows = state.range(0) / num_cols / sizeof(CType);
+  arrow::random::RandomArrayGenerator gen_{42};
+
+  std::vector<std::shared_ptr<Field>> fields = {};
+  std::vector<std::shared_ptr<Array>> columns = {};
+
+  for (int64_t i = 0; i < num_cols; ++i) {
+    fields.push_back(field("f" + std::to_string(i), ty));
+    columns.push_back(gen_.ArrayOf(ty, num_rows));
+  }
+  auto schema = std::make_shared<Schema>(std::move(fields));
+  auto batch = RecordBatch::Make(schema, num_rows, columns);
+
+  for (auto _ : state) {
+    ASSERT_OK_AND_ASSIGN(auto tensor, batch->ToTensor());
+  }
+  state.SetItemsProcessed(state.iterations() * num_rows * num_cols);
+  state.SetBytesProcessed(state.iterations() * ty->byte_width() * num_rows * num_cols);
+}
+
+void SetArgs(benchmark::internal::Benchmark* bench) {
+  for (int64_t size : {kL1Size, kL2Size}) {
+    for (int64_t number_of_columns : {3, 30, 300}) {
+      bench->Args({size, number_of_columns});
```

Comment:
Just a nit here, can we also give the parameters names to make the output more readable?
For example:
```suggestion
      bench->ArgNames({"size", "num_columns"});
```

Reply:
Sure! Just to be clear, do we need `bench->Args(...)` and `bench->ArgNames(...)` together or is one (`bench->ArgNames(...)`) enough?

Reply:
OK I see, I need to keep them both =)

Reply:
Nice =)

```
Running /var/folders/gw/q7wqd4tx18n_9t4kbkd0bj1m0000gn/T/arrow-archery-cow92p3n/WORKSPACE/build/release/arrow-tensor-benchmark
Run on (8 X 24 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB
  L1 Instruction 128 KiB
  L2 Unified 4096 KiB (x8)
Load Average: 23.41, 23.22, 14.54
----------------------------------------------------------------------------------------------------------------------
Benchmark                                                            Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------------------
BatchToTensorSimple<Int8Type>/size:65536/num_columns:3            2486 ns         2485 ns       302338 bytes_per_second=24.5587Gi/s items_per_second=26.3697G/s
BatchToTensorSimple<Int8Type>/size:65536/num_columns:30           3219 ns         3218 ns       217660 bytes_per_second=18.9628Gi/s items_per_second=20.3612G/s
BatchToTensorSimple<Int8Type>/size:65536/num_columns:300         13814 ns        13810 ns        49202 bytes_per_second=4.4103Gi/s items_per_second=4.73553G/s
BatchToTensorSimple<Int8Type>/size:4194304/num_columns:3        297191 ns       297041 ns         2401 bytes_per_second=13.1505Gi/s items_per_second=14.1203G/s
...
```